core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java

```diff
@@ -121,16 +120,17 @@ public BasicInvocationStage visitGetAllCommand(InvocationContext ctx, GetAllComm
                acquireLocalLocks(ctx, command, keys);
                stage = invokeNext(ctx, command);
             } else {
+               // Acquire the remote locks first, then the local locks
                final TxInvocationContext txContext = (TxInvocationContext) ctx;
                LockControlCommand lcc = cf.buildLockControlCommand(keys, command.getFlagsBitSet(),
                      txContext.getGlobalTransaction());
-               stage = invokeNext(ctx, lcc).thenCompose((stage1, rCtx, rCommand, rv) -> {
-                  acquireLocalLocks(rCtx, (FlagAffectedCommand) rCommand, keys);
+               stage = invokeNextThenApply(ctx, lcc, (rCtx, rLockCommand, rv) -> {
+                  acquireLocalLocks(rCtx, command, keys);
                   return invokeNext(rCtx, command);
                });
             }
          }
-         return stage.handle((rCtx, rCommand, rv, t) -> {
+         return makeStage(stage).andFinally(ctx, command, (rCtx, rCommand, rv, t) -> {
```

Comment:
Yeah, I guess I tend to use `andFinally` too much because don't like explicitly rethrowing the exception :)